apisix/plugins/degraphql.lua

```diff
@@ -0,0 +1,160 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core = require("apisix.core")
+local gq_parse = require("graphql").parse
+local req_set_body_data = ngx.req.set_body_data
+local ipairs = ipairs
+local pcall = pcall
+local type = type
+
+
+local schema = {
+    type = "object",
+    properties = {
+        query = {
+            type = "string",
+            minLength = 1,
+            maxLength = 1024,
+        },
+        variables = {
+            type = "array",
+            items = {
+                type = "string"
+            },
+            minItems = 1,
+        },
+        operation_name = {
+            type = "string",
+            minLength = 1,
+            maxLength = 1024
+        },
+    },
+    required = {"query"},
+}
+
+local plugin_name = "degraphql"
```

Comment:
I have a small question, should we call it `graphql-transcode` according to the `grpc` class plugin specification (`rest2grpc` => `grpc-transcode`) 🤔

Reply:
We borrow the name degraphql from Kong. What do you think about it, @juzhiyuan?

Reply:
From my understanding, `de` means `decode`, so `degraphql` means `decode GraphQL`. I think this name works because it does not belong to Kong only :)

Reply:
IMHO, it would be better to keep the original name, so we don't need to modify this PR and other things which are not public.

Reply:
+1